src/outputs/ascent.cpp

```diff
@@ -0,0 +1,203 @@
+    Node &mesh = root[meshblock_name];
+
+    // add basic state info
+    mesh["state/domain_id"] = pmb->gid;
+    mesh["state/cycle"] = tm->ncycle;
+    mesh["state/time"] = tm->time;
+
+    auto &bounds = pmb->cellbounds;
+    auto ib = bounds.GetBoundsI(IndexDomain::entire);
+    auto jb = bounds.GetBoundsJ(IndexDomain::entire);
+    auto kb = bounds.GetBoundsK(IndexDomain::entire);
+    int nx = ib.e - ib.s + 1;
+    int ny = jb.e - jb.s + 1;
+    int nz = kb.e - kb.s + 1;
+    uint64_t ncells = nx * ny * nz;
+
+    auto ib_int = bounds.GetBoundsI(IndexDomain::interior);
+    auto jb_int = bounds.GetBoundsJ(IndexDomain::interior);
+    auto kb_int = bounds.GetBoundsK(IndexDomain::interior);
+
+    auto &coords = pmb->coords;
+    Real dx1 = coords.CellWidth<X1DIR>(ib.s, jb.s, kb.s);
+    Real dx2 = coords.CellWidth<X2DIR>(ib.s, jb.s, kb.s);
+    Real dx3 = coords.CellWidth<X3DIR>(ib.s, jb.s, kb.s);
+    std::array<Real, 3> corner = coords.GetXmin();
+
+    // create the coordinate set
+    mesh["coordsets/coords/type"] = "uniform";
+
+    mesh["coordsets/coords/dims/i"] = nx + 1;
+    mesh["coordsets/coords/dims/j"] = ny + 1;
+    if (nz > 1) {
+      mesh["coordsets/coords/dims/k"] = nz + 1;
+    }
+
+    // add origin and spacing to the coordset (optional)
+    mesh["coordsets/coords/origin/x"] = corner[0];
+    mesh["coordsets/coords/origin/y"] = corner[1];
+    if (nz > 1) {
+      mesh["coordsets/coords/origin/z"] = corner[2];
+    }
+
+    mesh["coordsets/coords/spacing/dx"] = dx1;
+    mesh["coordsets/coords/spacing/dy"] = dx2;
+    if (nz > 1) {
+      mesh["coordsets/coords/spacing/dz"] = dx3;
+    }
+
+    // add the topology
+    mesh["topologies/topo/type"] = "uniform";
+    mesh["topologies/topo/coordset"] = "coords";
+
+    // indicate ghost zones with ascent_ghosts set to 1
+    Node &n_field = mesh["fields/ascent_ghosts"];
+    n_field["association"] = "element";
+    n_field["topology"] = "topo";
+    n_field["values"].set(conduit::DataType::int32(ncells)); // NOLINT
+    conduit::int32_array vals_array = n_field["values"].value();
+
+    int idx = 0;
+    for (int k = kb.s; k <= kb.e; k++) {
+      for (int j = jb.s; j <= jb.e; j++) {
+        for (int i = ib.s; i <= ib.e; i++) {
+          if ((i < ib_int.s) || (ib_int.e < i) || (j < jb_int.s) || (jb_int.e < j) ||
+              ((nz > 1) && ((k < kb_int.s) || (kb_int.e < k)))) {
+            vals_array[idx] = 1;
+          } else {
+            vals_array[idx] = 0;
+          }
+          idx++;
+        }
+      }
+    }
+
+    // create a field for each component of each variable pack
+    auto &mbd = pmb->meshblock_data.Get();
+
+    for (const auto &var : mbd->GetCellVariableVector()) {
+      PARTHENON_REQUIRE(var->GetDim(6) * var->GetDim(5) == 1,
+                        "Only scalar and vector fields are currently exported in Ascent. "
+                        "Please open an issue on GitHub if you have a use case.");
```

Comment:
Is there a deep reason for this? I haven't tried to ascent, but I don't see a reason not to at least simply flatten tensor fields and treat each component as a scalar? That would make it at least *possible* to visualize tonsorial quantities in a clunky way?

Reply:
The main reason is that (similar to prolong/restrict or outputs) that it'd be an implementation that "should" be correct without a proper test case.
Therefore, I opted for a loud fail so that once we have a use (and test case), this can be easily spotted an easily fixed. Otherwise, it could be another silent place where the code doesn't fail but also doesn't work correctly (because it's not properly tested).
Does this sound reasonable?

Reply:
I think the prolong/restrict and outputs with tensors has actually bitten us, as we've added many iterative ad hoc fixes before settling. So I don't see a reason to wait.

The use-case I had in mind was, e.g., metric or radiation quantities in Phoebus. I don't think it would be too cumbersome to throw in a derived field in advection or something with tensorial shape that gets flattened by ascent. But seeing as I haven't tried ascent with Phoebus, it doesn't need to be prioritized right now.

Actually I just realized something: This will cause phoebus to crash if compiled with ascent support because it checks EVERY variable, not just the ones registered with ascent. So at the *very* least this should *skip* tensorial fields, rather than dying.

Reply:
This restriction is now removed (by using component labels for a flat index).